### ISSUE TITLE
add electric field including vE^2 term

### ISSUE THREAD

**Issue by krystophny** [itpplasma/SIMPLE#71]

**Comment by krystophny** [comment]
*Automated triage pass (Claude), grounded in the repository at `d8b0e01`. Verify before acting.*

## Purpose

The canonical/symplectic guiding-centre Hamiltonian in SIMPLE is purely magnetic. `src/field_can.f90:261`:

```fortran
f%H = f%vpar**2/2d0 + f%mu*f%Bmod
```

and the surviving 2018 placeholder at `src/field_can.f90:316`:

```fortran
f%d2H(1:6) = f%vpar*f%d2vpar(1:6) + f%mu*f%d2Bmod ! + qi*d2Phie
```

That comment was introduced by `fef898e` ("working version symplectic", C. Albert, 2018-10-25), then in `SRC/field_can.f90`; `git log --oneline -S'Phie' -- src/` only shows `0a624fa` ("Move to src") because of the path rename.

A radial electric field matters for alpha-particle confinement in two ways. The E×B rotation shifts the toroidal precession of trapped particles, which moves the boundary between confined and stochastically lost orbits. And where `v_E` becomes a non-negligible fraction of the residual speed, the `v_E²` term in the guiding-centre Hamiltonian is no longer negligible relative to `μB`. The issue title asks for both: `Φ` *and* the second-order `v_E²` term.

## Current state — read this before planning

**There is already a partial, live electrostatic implementation on the non-symplectic path.** Any spec that ignores it will create a second, inconsistent mechanism.

`src/sub_alpha_lifetime_can.f90:12-29` defines a stub:

```fortran
subroutine elefie_can(x,derphi)
    ! Computes the derivatives of the electrostatic potential over
    ! coordinates (covariant electric field). Potential is normalized
    ! to reference temperature : phinor = e phi / T.
    real(dp), intent(in) :: x(3)
    real(dp), intent(out) :: derphi(3)
    derphi=0.d0
end subroutine elefie_can
```

`velo_can` (same file, lines 33-159) already consumes it with the full **first-order** electrostatic physics: the E×B drift `a_phi(1:3)` (lines 124-128), the parallel-energy change `phidot` / `vz(4)` (lines 148, 154), and the pitch equation `vz(5)` (lines 156-157). This is live code, not a dead branch — `velo_can` is called from `src/spectre_orbit.f90:256,282`, and `orbit_timestep_axis` from the same module is used by `src/simple_main.f90:1600,1871`, `src/classification.f90:71`, `src/simple.f90:284` and several tests. The file is in the build (`src/CMakeLists.txt:33`).

So the accurate statement of the gap is:

| Path | Φ term | v_E² term |
|---|---|---|
| RK / `velo_can` (`src/sub_alpha_lifetime_can.f90`) | machinery present, source stubbed to zero | absent |
| Symplectic / canonical (`src/field_can.f90`) | absent | absent |

Neither path has an input surface: the `&config` namelist (`src/params.f90:164-188`) and the `&profiles` namelist (`src/profiles.f90:36-40`, which carries `Te/Ti/ni` only) have no potential key.

Note the normalization already in the tree: `phinor = e·φ/T`. A new implementation should either reuse it or explicitly state why the canonical path needs a different one.

The Hamiltonian lives in three procedures in `src/field_can.f90`: `get_val` (line 251), `get_derivatives` (line 269), `get_derivatives2` (line 294). Every integrator — `src/orbit_symplectic.f90`, `src/orbit_symplectic_quasi.f90`, `src/orbit_rk_core.f90` — consumes only `f%H, f%dH, f%d2H, f%pth, f%dpth, f%d2pth, f%vpar`. The key structural fact: **both new terms depend on `(r, θ, φ)` only, never on `pphi`**, so `f%dH(4) = f%vpar/f%hph`, `f%dpth` and `f%d2H(7:9)` are unchanged, the symplectic structure is untouched, and every integrator variant stays symplectic. The change is confined to the Hamiltonian.

### The metric problem — this is the real blocker

`field_can_t` (`src/field/field_can_base.f90:12-34`) has no slot for `Φ` and no metric. `v_E² = c²|∇_⊥Φ|²/B²` needs `g^{ij}`, which the canonical representation deliberately does not carry.

`ref_coords` (`src/coordinates/reference_coordinates.f90:13`) is a `class(coordinate_system_t), allocatable`, and `coordinate_system_t` does declare `procedure(metric_tensor_if), deferred :: metric_tensor` (libneo `src/coordinates/libneo_coordinates_base.f90:65`, interface at lines 93-98), implemented for chartmaps by `chartmap_metric_tensor` (libneo `src/coordinates/libneo_coordinates_chartmap.f90`, bound at line 37, defined at 554-586). libneo is a fetched dependency, not vendored in this checkout — read it under `build/_deps/libneo-src/`.

Two constraints follow, both of which must shape the plan:

1. **`metric_tensor` returns `g`, `ginv`, `sqrtg` and no derivatives of them.** The symplectic integrators call the evaluator with `mode_secders = 2` (e.g. `src/orbit_symplectic.f90:345,377`), so they need `d2H(1:6)`. Getting `d(vE2)` and `d2(vE2)` therefore requires derivatives of `ginv` that libneo does not expose. There is no way around this except finite-differencing `metric_tensor` or extending libneo.
2. **`ref_coords` is not available for every field.** `init_reference_coordinates` (`src/coordinates/reference_coordinates.f90:18-60`) requires a non-empty `coord_input` naming a chartmap or a VMEC wout and `error stop`s otherwise. `metric_tensor` is currently called nowhere in `src/` — this would be its first use in SIMPLE.

This is why the work must be staged: **Φ first (no metric needed), v_E² second (metric needed).**

## Reproduction / evidence
There is nothing to reproduce; the feature is absent. The verification recipe is in the scenarios below. Build and test commands, confirmed against `Makefile` and `CMakeLists.txt:64-66`:

```bash
make CONFIG=Fast                        # "Fast" is a valid build type
make test TEST=test_electric_field      # Makefile:24 passes TEST to ctest -R
make test TEST=test_sympl_testfield
make test-golden-fast                   # Makefile:105
make test-regression                    # Makefile:78
```

## Behaviour (BDD)

```gherkin
Scenario: Zero potential is a no-op
  Given test/golden_record/boozer/simple.in unchanged, with no &electric namelist
  When `make test-golden-fast` runs
  Then confined_fraction.dat and times_lost.dat match the stored golden records
   with the existing tolerances

Scenario: Every existing input still parses
  Given every simple.in under test/golden_record/
  When each is read by read_electric_config
  Then no error is raised and phi_scale is left at its default 0d0

Scenario: Potential and its derivatives are exact for a power series
  Given potential_type="power_series", phi_scale=1d-3, phi_coef=[0,0,1,0,0,0,0,0,0,0]
    And therefore Phibar(s) = 1d-3 * s**2
  When eval_electric is called at s=0.36, th=0.7, ph=1.1 with mode_secders=2
  Then f%Phie equals 1d-3*0.1296 to 1d-14 relative
    And f%dPhie(1) equals 2*1d-3*0.36 to 1d-14 relative
    And f%dPhie(2) and f%dPhie(3) are exactly 0d0
    And f%d2Phie(1) equals 2*1d-3 to 1d-14 relative
    And f%d2Phie(2:6) are exactly 0d0

Scenario: The Hamiltonian derivatives stay consistent with H
  Given the analytic field of src/field/field_can_test.f90 (isw_field_type=TEST)
    And Phibar(s) = 1d-3 * s**2, include_vE2 = .false.
  When get_derivatives2 is evaluated at r=0.3, th=0.7, ph=1.1, pphi=0.05
  Then f%dH(1:3) matches central finite differences of f%H with step 1d-5
   to 1d-7 relative
    And f%d2H(1:6) matches central finite differences of f%dH(1:3) with step 1d-5
   to 1d-5 relative
    And f%dH(4), f%dpth(1:4) and f%d2pth(1:10) are bit-identical to the values
   obtained with phi_scale = 0d0

Scenario: Energy is conserved with a potential switched on
  Given isw_field_type=TEST, integmode=3, orbit_model=0 (ORBIT_GC, src/params.f90:73)
    And ntimstep chosen so that dtau/dtaumin is an integer
     (src/simple.f90:136 error-stops otherwise)
    And Phibar(s) = 1d-2 * s**2, include_vE2 = .false.
  When one trapped particle is traced for 1d4 normalized time units
  Then max|H(t) - H(0)|/|H(0)| < 1d-6 with no secular drift

Scenario: vE2 kernel is correct against a hand-supplied inverse metric
  Given the pure kernel vE2_kernel(dPhi, ginv, hcov, Bmod)
    And ginv = diag(1, 4, 9), hcov = [0, 0.2, 0.8], Bmod = 1.5
    And dPhi = [0.1, 0.3, 0.5]
  When vE2_kernel is called
  Then it returns (ro0/Bmod)**2 * (g_ij dPhi_i dPhi_j - (g_ij hcov_i dPhi_j)**2)
   evaluated in double precision, to 1d-13 relative
   i.e. the test asserts the formula independently of any coordinate system

Scenario: vE2 derivatives match finite differences of vE2 itself
  Given isw_field_type=BOOZER with a chartmap coord_input, include_vE2 = .true.
  When eval_electric is called with mode_secders=2 at an interior point
  Then f%dvE2 matches central differences of f%vE2 with step 1d-5 to 1d-6 relative
    And f%d2vE2 matches central differences of f%dvE2 to 1d-4 relative

Scenario: Unsupported combinations fail loudly, never silently
  Given phi_scale /= 0d0
  When isw_field_type is MEISS, ALBERT, CANFLUX or SPECTRE
  Then the run stops with a message naming the field type and the &electric namelist
  And when include_vE2 is .true. and ref_coords is not allocated
  Then the run stops with a message saying vE2 requires a reference coordinate system
```

## Implementation plan

### Stage 1 — potential only, no metric (one PR)

1. **`src/field/field_can_base.f90`** — extend `field_can_t` (declared lines 12-34), after the `d2Bmod` block:
   ```fortran
   real(dp) :: Phie = 0d0
   real(dp), dimension(3) :: dPhie = 0d0
   real(dp), dimension(6) :: d2Phie = 0d0
   real(dp) :: vE2 = 0d0
   real(dp), dimension(3) :: dvE2 = 0d0
   real(dp), dimension(6) :: d2vE2 = 0d0
   ```
   Default initializers make every existing evaluator a no-op. Check that this does not break the `!$acc routine seq` / `!$omp threadprivate` usage of the type (`src/simple_gpu.f90`); if it does, zero the fields explicitly in `field_can_init` instead.

2. **New file `src/field_electric.f90`**, module `field_electric`:
   - `&electric` namelist: `character(len=32) :: potential_type = "none"`, `real(dp) :: phi_scale = 0d0`, `real(dp) :: phi_coef(10) = 0d0` (power series in `s`, mirroring the `*_coef` arrays at `src/profiles.f90:19-28`), `logical :: include_vE2 = .false.` (stage 1 default).
   - `read_electric_config(config_file)` modelled on `read_profiles` (`src/profiles.f90:44-60`): open, `read(unit, nml=electric, iostat=ios)`, ignore `ios`, close. An absent namelist leaves every default in place, so all existing `simple.in` files stay valid.
   - `electric_normalization(...)` — decide and **document** the convention. Two options: reuse the existing `phinor = e·φ/T` of `elefie_can`, or define the internal-unit form. If the latter: from `src/params.f90:288-290` (`v0 = sqrt(2 E_alpha ev/(n_d m_p))`, `ro0 = v0 n_d m_p c/(n_e e)`) and `src/simple.f90:145-146` (`f%ro0 = ro0/sqrt(2)`, `f%vpar = z0(4)*z0(5)*sqrt(2)`), the internal velocity unit is `v_ref = v0/sqrt(2)`, giving `Phibar = e·Φ/(n_d m_p v_ref²)`. **Derive this once in a comment in the source and unit-test it against a hand-computed number** rather than trusting this paragraph.
   - `eval_electric(f, r, th_c, ph_c, mode_secders)` filling `f%Phie/dPhie/d2Phie` from the power series in `s = r`. No metric is involved; all derivatives are exact.
   - `error stop` with a message naming the field type if `phi_scale /= 0d0` and `isw_field_type` is `MEISS`, `ALBERT`, `CANFLUX` or `SPECTRE` (constants in `magfie_sub`, used at `src/field_can.f90:44-78`).

3. **`src/field_can.f90`** — add a concrete wrapper beside the `evaluate` pointer (declared at line 25):
   ```fortran
   subroutine eval_field(f, r, th_c, ph_c, mode_secders)
     type(field_can_t), intent(inout) :: f
     real(dp), intent(in) :: r, th_c, ph_c
     integer, intent(in) :: mode_secders
     call evaluate(f, r, th_c, ph_c, mode_secders)
     call eval_electric(f, r, th_c, ph_c, mode_secders)
   end subroutine eval_field
   ```
   Extend `get_val` (line 251) with `+ f%Phie - 0.5d0*f%vE2`, `get_derivatives` (line 269) `f%dH(1:3)` with `+ f%dPhie - 0.5d0*f%dvE2`, and `get_derivatives2` (line 294) `f%d2H(1:6)` with `+ f%d2Phie - 0.5d0*f%d2vE2`, replacing the stale comment at line 316. **Do not touch `f%dH(4)`, `f%dpth`, `f%d2pth`, or `f%d2H(7:9)`.**

4. **Retarget all ten `eval_field => evaluate` aliases.** The comment that only three exist is wrong; `rg -n 'eval_field => evaluate' src/` returns exactly ten sites:

   | File | Line |
   |---|---|
   | `src/simple.f90` | 12 |
   | `src/params.f90` | 13 |
   | `src/orbit_symplectic.f90` | 6 |
   | `src/orbit_symplectic_base.f90` | 2 |
   | `src/orbit_symplectic_quasi.f90` | 4 |
   | `src/orbit_rk_core.f90` | 16 |
   | `src/orbit_fo_field.f90` | 24 |
   | `src/spectre_sympl_orbit.f90` | 18 |
   | `src/diag/diag_newton.f90` | 8 |
   | `src/diag/diag_orbit.f90` | 9 |

   In each, change `eval_field => evaluate` to a plain `eval_field`. Missing any one silently drops the potential in that code path — this is the single largest regression risk in the change. `src/simple_gpu.f90:83,176` and `src/field/field_can_boozer.f90:38` call `eval_field_booz` directly and bypass the wrapper: decide explicitly whether the GPU path is in scope, and if not, `error stop` when a potential is configured together with the GPU path.

5. **`src/simple_main.f90`** — call `read_electric_config(config_file)` next to the existing `read_profiles_config(config_file)` at line 146, and `electric_normalization(...)` after `E_alpha`, `n_e`, `n_d` are set.

6. **`src/CMakeLists.txt`** — add `field_electric.f90` to `SOURCES` (the `simple` static library, `add_library (simple STATIC ${SOURCES})` at line 94), next to `field_can.f90` (line 28).

7. **Tests** — new `test/tests/test_electric_field.f90` covering scenarios 3, 4 and 6, registered in `test/tests/CMakeLists.txt` following the `test_orbit_frequencies` pattern at lines 353-358 (simpler than the `test_sympl_testfield` block at 341-352, which only needs `-cpp` for `WOUT_FILE`).

8. **Docs** — add a bullet to `DOC/config.md` describing the `&electric` namelist and stating the chosen normalization explicitly.

### Stage 2 — the v_E² term (separate PR)

9. Add `vE2_kernel(dPhi, ginv, hcov, Bmod, ro0)` as a **pure function** in `src/field_electric.f90` so it is unit-testable without any coordinate system:
   `vE2 = (ro0/Bmod)**2 * (g^{ij} ∂_i Φ ∂_j Φ − (g^{ij} h_i ∂_j Φ)²)`
   with `h = (0, f%hth, f%hph)`. `h_r = 0` is structural, not an assumption: `field_can_t` has no `hr` component at all (`src/field/field_can_base.f90:13`). Sign convention: the `−v_E²/2` term belongs to the weak-flow drift ordering in which the symplectic 1-form is left as `A + ρ₀ v_∥ h`, which is what SIMPLE does. Confirm it numerically before merging.

10. Obtain `ginv` from `ref_coords%metric_tensor(u, g, ginv, sqrtg)`, pushing through the `integ_to_ref` Jacobian. For the Boozer chartmap that Jacobian is diagonal (`xref(1) = sqrt(xinteg(1))`, `src/field/field_can_boozer.f90:11-21`).

11. **Derivatives of `ginv` do not exist in the libneo API.** Pick one and document it: (a) central-difference `metric_tensor` on the reference grid to build `dvE2` and `d2vE2`, accepting the accuracy cost, or (b) add a `metric_tensor_derivatives` binding to libneo. Option (a) keeps the change inside SIMPLE; option (b) is a libneo PR. Do not assume `metric_tensor` alone is sufficient.

12. Optional: the electrostatic half-kick in `fo_step` (`src/orbit_fo_boris.f90:444-471`), `v = v + 0.5*dt*qcm_E*E_cart` before and after the magnetic rotation, reusing the Jacobian already assembled in `cart_field` (called at line 454). This enables a GC/FO cross-check with `orbit_model = 7` (`ORBIT_FULL_ORBIT`, `src/params.f90:74`).

13. Decide what to do about `elefie_can` (`src/sub_alpha_lifetime_can.f90:12`). Either wire it to the same `&electric` input so the RK and symplectic paths agree, or leave it stubbed and document that the potential is symplectic-path-only. Leaving it silently at zero while the canonical path has a potential means the two orbit models disagree, which will surface as a confusing test failure later.

## Acceptance criteria

- [ ] `make CONFIG=Fast` builds clean with no new warnings.
- [ ] `make test-golden-fast` and `make test-regression` reproduce all stored golden records unchanged when no `&electric` namelist is present.
- [ ] Every `simple.in` under `test/golden_record/*/` still parses unmodified.
- [ ] `make test TEST=test_electric_field` passes with the tolerances given in scenarios 3, 4 and 6.
- [ ] Energy drift `max|H(t)-H(0)|/|H(0)| < 1e-6` over 1e4 normalized time units with `phi_scale` nonzero at `integmode=3`.
- [ ] `rg -n 'eval_field => evaluate' src/` returns **zero** matches. (Note: `rg -n '=> evaluate' src/` will still match the seven dispatch assignments in `src/field_can.f90` lines 6, 46, 50, 56, 65, 73, 77, which are correct and must stay — do not use that looser pattern as a gate.)
- [ ] Requesting `phi_scale /= 0d0` with `isw_field_type` in {meiss, albert, canflux, spectre}, or `include_vE2 = .true.` without an allocated `ref_coords`, fails with a named error rather than silently ignoring the field.
- [ ] The chosen normalization is stated in `DOC/config.md` and asserted by at least one unit test against a hand-computed value.

## Scope and risk

Two PRs, staged by whether the metric is needed: (1) the `Φ` term plus `&electric` input and unit tests, restricted to Boozer/test fields — size/M, tier T2, it touches the core Hamiltonian; (2) the `v_E²` term, which needs metric derivatives that libneo does not currently expose, plus the optional Boris E-kick — size/M, and its cost depends on which of the two options in step 11 is chosen.

Main risks, in order:

1. **Missing one of the ten alias sites**, silently dropping the potential in that code path. Mitigated by the `rg` gate above.
2. **Metric derivatives** (step 11) — this is the item most likely to be underestimated; scope it before starting stage 2.
3. **Divergence from the existing `elefie_can` path** (step 13), giving two orbit models that disagree about the same physical input.
4. Silent regression in the magnetic-only case, caught by the golden records.

## References

- Littlejohn, R. G. (1983), *Variational principles of guiding centre motion*, J. Plasma Phys. **29**, 111 — the guiding-centre Lagrangian SIMPLE's canonical form derives from.
- Brizard, A. J. and Hahm, T. S. (2007), *Foundations of nonlinear gyrokinetic theory*, Rev. Mod. Phys. **79**, 421 — standard reference for the second-order guiding-centre Hamiltonian in the representation where the symplectic 1-form is left unmodified. Check the sign of the `v_E²` term against this rather than against the summary above.
- White, R. B. and Chance, M. S. (1984), Phys. Fluids **27**, 2455 — ORBIT: Hamiltonian guiding-centre in Boozer coordinates with an electrostatic potential; the closest structural analogue.
- Albert, C. G., Kasilov, S. V., Kernbichler, W. (2020), J. Comput. Phys. **403**, 109065 — SIMPLE's symplectic integrators; the Hamiltonian being modified here is the one in that paper.
- Hirvijoki, E. et al. (2014), Comput. Phys. Commun. **185**, 1310 — ASCOT; a possible cross-code comparison, though the exact form of its guiding-centre electrostatic term was not checked here.
- Landreman, M. et al. (2014), Phys. Plasmas **21**, 042503 — SFINCS-related; relevant to how `E_r` enters stellarator neoclassical transport.

The choice between the weak-flow (`−v_E²/2` in `H`, unmodified 1-form) and strong-flow (`+v_E²/2` in `H` plus `+m v_E` in the 1-form) formulations is an implementation decision internal to this code; the strong-flow variant would break the canonical structure and is out of scope. No specific literature reference is asserted here for that split beyond the Brizard–Hahm review above.
